airbyte-ci/connectors/pipelines/Python connector images analysis.md

```diff
@@ -0,0 +1,738 @@
+            images_with_custom_pkgs.add(connector_image)
+
+
+```
+
+    airbyte/source-instagram:1.0.11 is installing custom system packages: libcrypt-dev:arm64, libtirpc-dev:arm64, libc6-dev:arm64, rpcsvc-proto, libnsl-dev:arm64, linux-libc-dev:arm64, libc-dev-bin
+    airbyte/source-twilio:0.10.0 is installing custom system packages: libcrypt-dev:arm64, libtirpc-dev:arm64, libc6-dev:arm64, rpcsvc-proto, libnsl-dev:arm64, linux-libc-dev:arm64, libc-dev-bin
+    airbyte/source-gitlab:1.6.0 is installing custom system packages: libcrypt-dev:arm64, libtirpc-dev:arm64, libc6-dev:arm64, rpcsvc-proto, libnsl-dev:arm64, linux-libc-dev:arm64, libc-dev-bin
+    airbyte/source-google-analytics-v4:0.2.1 is installing custom system packages: libcrypt-dev:arm64, libtirpc-dev:arm64, libc6-dev:arm64, rpcsvc-proto, libnsl-dev:arm64, linux-libc-dev:arm64, libc-dev-bin
+    airbyte/source-salesforce:2.1.4 is installing custom system packages: libcrypt-dev:arm64, libtirpc-dev:arm64, libc6-dev:arm64, rpcsvc-proto, libnsl-dev:arm64, linux-libc-dev:arm64, libc-dev-bin
+
+
+**The images listed above have extra system packages compared to there base image. But no instruction in the dockerfile installs these system packages. My current hypothesis is that the base image they use had these system packages in previous versions. Let's rebuild these connector and see if the "custom packages" are still there.**
+
+
+```python
+async with dagger.Connection(dagger.Config(log_output=sys.stderr)) as dagger_client:
+    for connector_image in images_with_custom_pkgs:
+        connector = Connector(connector_image.split(":")[0].replace("airbyte/", ""))
+        connector_container = dagger_client.host().directory(str(connector.code_directory)).docker_build()
+        connector_container_pkgs = await get_installed_system_packages(connector_container, dagger_client)
+        custom_pkgs = connector_container_pkgs - base_pkgs
+        assert not custom_pkgs
+        print(f"No custom packages found after local build of {connector}")
+```
+
+    No custom packages found after local build of source-instagram
+    No custom packages found after local build of source-twilio
+    No custom packages found after local build of source-gitlab
+    No custom packages found after local build of source-google-analytics-v4
+    No custom packages found after local build of source-salesforce
+
+
+#### Takeaways for the future base image
+
+1. We don't have any certified python connector that is installing a custom system dependency
+2. The use of base images tag that can change the underlying image in use (like python3.9-slim: it might get updated on each patch version of python 3.9) proves again that we can't achieve reproductible build with these images. We should definitely be targetting tag with their `sha256`: e.g `python:3.9@sha256:0596c508fdfdf28fd3b98e170f7e3d4708d01df6e6d4bffa981fd6dd22dbd1a5`. This will ensure that on rebuild the same base image will be used.
```

Comment:
I agree. What's our proccess for deciding to upgrade to a different SHA? Are we going to scan this image with `snyk` or another package to keep up with any potential security issues (maybe out of scope and something that this project rather enables, but thinking forward)

Reply:
The process is to cut a new base image version that will use a different base image. This is something easier to grasp by[ checking this out](https://github.com/airbytehq/airbyte/pull/30303/files#diff-72e55cebc5c066e9ae9beb2a06a516d66ed66dc7e2c4740514b3fc740d2541baR53): it would mean changing the value of the `base_base_image` attribute here.

Reply:
Cool - I more meant to ask what our process for deciding to upgrade was, versus how we would upgrade. Is this something we plan to try to do on a regular schedule? Something we would only change if we notice issues with our image? Something we'd do manually, or automate (e.g. with a dependency bot or some other thing that would change it for us and make sure that the tests still pass)?

Reply:
I think it's currently *something we change if we notice issues with our image.*
For your `snyk` example:
If a vulnerability is discovered:
- We check if  Python released a patched image, and cut a new version our our base image which is using  the patched Python version as the base image.
- we patch the vulnerabilites ourself otherwise (so defining our own base image is definitely an enabler in this context), and still cut a new version.

Reply:
> I think it's currently something we change if we notice issues with our image.

sounds good! As long as we've made a decision, good with me. I'd love to make sure we have a(n issue for implementing a) process for being aware of any issues with the image we choose (e.g. by using snyk)